File: Src/MusicPlayerBackend/App/MinioExtensions.cs
using Minio;
using Minio.DataModel.Args;

namespace MusicPlayerBackend.App;

public static class MinioExtensions
{
    public static bool BucketExists(this IMinioClient minioClient, BucketExistsArgs args)
    {
        return minioClient.BucketExistsAsync(args).ConfigureAwait(false).GetAwaiter().GetResult();
    }

    public static IMinioClient MakeBucket(this IMinioClient minioClient, MakeBucketArgs args)
    {
        minioClient.MakeBucketAsync(args).ConfigureAwait(false).GetAwaiter().GetResult();

        return minioClient;
    }

    public static IMinioClient CreateBucketIfNotExists(this IMinioClient minioClient, string name)
    {
        if (!minioClient.BucketExists(new BucketExistsArgs().WithBucket(name)))
            minioClient.MakeBucket(new MakeBucketArgs().WithBucket(name));

        return minioClient;
    }
}

File: Src/Host/MinioExtensions.fs
namespace MusicPlayerBackend.Host.Ext

open System.Runtime.CompilerServices
open Minio
open Minio.DataModel.Args

type MinioExtensions() =

    [<Extension>]
    static member BucketExists(mc: IMinioClient, args: BucketExistsArgs) =
        mc.BucketExistsAsync(args).ConfigureAwait(false).GetAwaiter().GetResult()

    [<Extension>]
    static member MakeBucket(mc: IMinioClient, args: MakeBucketArgs) =
        mc.MakeBucketAsync(args).ConfigureAwait(false).GetAwaiter().GetResult()
        mc

    [<Extension>]
    static member CreateBucketIfNotExists(mc: IMinioClient, name) =
        if mc.BucketExists(BucketExistsArgs().WithBucket(name)) = false then
            mc.MakeBucket(MakeBucketArgs().WithBucket(name))
        else
            mc
